datafusion/core/src/datasource/datasource.rs

```diff
@@ -98,12 +100,11 @@ pub trait TableProvider: Sync + Send {
         None
     }
 
-    /// Insert into this table
-    async fn insert_into(
-        &self,
-        _state: &SessionState,
-        _input: Arc<dyn ExecutionPlan>,
-    ) -> Result<Arc<dyn ExecutionPlan>> {
+    /// Return a [`DataSink`] suitable for writing to this table
```

Comment:
Here is the new interface -- I think this probably should also get some sort of "options" parameter too so that we can provide a way to pass in settings like `row_group_size` from the various inserts). 

I prefer to add that parameter at some future point when I have an actual usecase with `COPY` rather than guessing exactly what is needed.